### PR TITLE
Cifar10 example

### PR DESCRIPTION
**Pull request by valeriano-manassero**
Following discussion : https://github.com/tensorflow/tfx/issues/572 , this is first draft with an Cifar10 example working on an on premise instance of Kubeflow.

The pipeline can be executed in the interactive version or built and pushed to a kubeflow scheduler.

I'm pretty sure that many improvements can be applied to this example, this PR is meant to have a starting point for a good example about image related pipeline.

### CODE REVIEW

File: tfx/examples/cifar10_pipeline/README.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+Note: This site provides applications using data that has been modified
+for use from its original source, The Canadian Institute for Advanced Research (CIFAR).
+The Canadian Institute for Advanced Research (CIFAR) makes no claims as to the content,
+accuracy, timeliness, or completeness of any of the data provided at this site.
+The data provided at this site is subject to change at any time. It is
+understood that the data provided at this site is being used at one’s own risk.
+
+You can read more about the dataset in [CIFAR dataset homepage](https://www.cs.toronto.edu/~kriz/cifar.html).
+
+## Run the example on Kubeflow (on premise)
```
[the remark this answers is not in the document]
Comment (neuromage) [reply tfx/examples/cifar10_pipeline/README.md]: It looks like persistent volumes are what's being used here? I believe it will work on GCP as well. So I think it's not specific to on-premise deployments only.

[the remark this answers is not in the document]
Comment (valeriano-manassero) [reply tfx/examples/cifar10_pipeline/README.md]: Ok, so I removed references to on prem in README. Should be ok now.

**Comment by robertlugg** [comment]
I'm happy to see the effort.  _At the moment_, it simply demonstrates that TFX works for TFRecord inputs with images, which is nice.  However, I didn't see anything being demonstrated beyond what's already in the existing demos.

When working with images, what does TFX provide you?  Maybe show me the distribution of labels.  Are they uniformly distributed?  As an example, the statististics:
tfdv
really aren't that interesting.  Perhaps more features or metrics need to be computed.  Maybe the min/max grey level of the colors, or check to see if any of the sizes are incorrect, or...

Basically, I love the idea of a TFX demo flow using images.  But I would suggest that the demo needs to show more than what can be seen with the taxicab demo.  I think your KubeFlow stuff is worth showing off.

**Comment by valeriano-manassero** [comment]
@robertlugg ty for your typo fix and your suggestions. I completely agree on what you said about improving this example. With this PR I just wanted to scaffold a pipeline implementation for images. I'm convinced that many improvements can be made. Unfortunately I don't have so much free time atm so I cannot say if and when I will improve this.

File: tfx/examples/cifar10_pipeline/cifar10_pipeline_kubeflow.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    validate_stats = ExampleValidator(
+        stats=statistics_gen.outputs.output,
+        schema=infer_schema.outputs.output)
+    transform = Transform(
+        input_data=example_gen.outputs.examples,
+        schema=infer_schema.outputs.output,
+        module_file=module_file)
+    trainer = Trainer(
+        module_file=module_file,
+        transformed_examples=transform.outputs.transformed_examples,
```
Comment (ruoyu90) [inline tfx/examples/cifar10_pipeline/cifar10_pipeline_kubeflow.py]: Use `examples` as `transformed_examples` will be deprecated.

File: tfx/examples/cifar10_pipeline/cifar10_pipeline_kubeflow.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+def _create_pipeline(pipeline_name: Text, pipeline_root: Text, data_root: Text,
+                     module_file: Text, serving_model_dir: Text) -> pipeline.Pipeline:
+    examples = external_input(data_root)
+    input_split = example_gen_pb2.Input(splits=[
+        example_gen_pb2.Input.Split(name='train', pattern='train.tfrecord'),
+        example_gen_pb2.Input.Split(name='eval', pattern='test.tfrecord')
+    ])
+    example_gen = ImportExampleGen(input_base=examples, input_config=input_split)
+    statistics_gen = StatisticsGen(input_data=example_gen.outputs.examples)
+    infer_schema = SchemaGen(stats=statistics_gen.outputs.output)
```
Comment (ruoyu90) [inline tfx/examples/cifar10_pipeline/cifar10_pipeline_kubeflow.py]: You can use infer_feature_shape=True so that you don't need to deal with VarlenFeature in Transform.

**Comment by valeriano-manassero** [comment]
After rebasing onto latest master commit I cannot build the pipeline anymore:
```Traceback (most recent call last):
  File "/home/ummon/Documents/github.com/valeriano-manassero/tfx/venv/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3326, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-833684f0c5a2>", line 1, in <module>
    runfile('/home/ummon/Documents/github.com/valeriano-manassero/tfx/tfx/examples/cifar10_pipeline/cifar10_pipeline_kubeflow.py', wdir='/home/ummon/Documents/github.com/valeriano-manassero/tfx/tfx/examples/cifar10_pipeline')
  File "/snap/pycharm-professional/154/helpers/pydev/_pydev_bundle/pydev_umd.py", line 197, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/snap/pycharm-professional/154/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/ummon/Documents/github.com/valeriano-manassero/tfx/tfx/examples/cifar10_pipeline/cifar10_pipeline_kubeflow.py", line 115, in <module>
    KubeflowRunner(config=config).run(_pipeline)
  File "/home/ummon/Documents/github.com/valeriano-manassero/tfx/tfx/orchestration/kubeflow/kubeflow_dag_runner.py", line 226, in run
    workflow = self._compiler.create_workflow(
AttributeError: 'Compiler' object has no attribute 'create_workflow'```

It's worth to open a new issue?

**Comment by numerology** [comment]
> After rebasing onto latest master commit I cannot build the pipeline anymore:
> 
> ```
>   File "/home/ummon/Documents/github.com/valeriano-manassero/tfx/venv/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 3326, in run_code
>     exec(code_obj, self.user_global_ns, self.user_ns)
>   File "<ipython-input-2-833684f0c5a2>", line 1, in <module>
>     runfile('/home/ummon/Documents/github.com/valeriano-manassero/tfx/tfx/examples/cifar10_pipeline/cifar10_pipeline_kubeflow.py', wdir='/home/ummon/Documents/github.com/valeriano-manassero/tfx/tfx/examples/cifar10_pipeline')
>   File "/snap/pycharm-professional/154/helpers/pydev/_pydev_bundle/pydev_umd.py", line 197, in runfile
>     pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
>   File "/snap/pycharm-professional/154/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
>     exec(compile(contents+"\n", file, 'exec'), glob, loc)
>   File "/home/ummon/Documents/github.com/valeriano-manassero/tfx/tfx/examples/cifar10_pipeline/cifar10_pipeline_kubeflow.py", line 115, in <module>
>     KubeflowRunner(config=config).run(_pipeline)
>   File "/home/ummon/Documents/github.com/valeriano-manassero/tfx/tfx/orchestration/kubeflow/kubeflow_dag_runner.py", line 226, in run
>     workflow = self._compiler.create_workflow(
> AttributeError: 'Compiler' object has no attribute 'create_workflow'```
> 
> It's worth to open a new issue?
> ```

Hi @valeriano-manassero , this is due to a recent change brought by KFP. Can you upgrade KFP to version 0.1.30? Thanks!

File: tfx/examples/cifar10_pipeline/cifar10_utils.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+
+def _get_raw_feature_spec(schema):
+    return schema_utils.schema_as_feature_spec(schema).feature_spec
+
+
+def _fill_in_missing(x):
+    default_value = '' if x.dtype == tf.string else 0
+    return tf.squeeze(
+        tf.sparse.to_dense(
```
Comment (ruoyu90) [inline tfx/examples/cifar10_pipeline/cifar10_utils.py]: Now that the schema contains shape info, we are directly reading DenseFeature instead of SparseFeature, so this is not necessary I think.

**Comment by robertlugg** [comment]
@googlebot I consent

File: tfx/examples/cifar10_pipeline/README.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+* [Trainer](https://github.com/tensorflow/tfx/blob/master/docs/guide/trainer.md)
+  trains the model using TensorFlow [Estimators](https://www.tensorflow.org/guide/estimators)
+* [Evaluator](https://github.com/tensorflow/tfx/blob/master/docs/guide/evaluator.md)
+  performs deep analysis of the training results.
+* [ModelValidator](https://github.com/tensorflow/tfx/blob/master/docs/guide/modelval.md)
+  ensures that the model is "good enough" to be pushed to production.
+* [Pusher](https://github.com/tensorflow/tfx/blob/master/docs/guide/pusher.md)
+  deploys the model to a serving infrastructure.
+
+## The dataset
```
Comment (ruoyu90) [inline tfx/examples/cifar10_pipeline/README.md]: Could you also add more documentation on how to get the training data in tfrecord format?

File: tfx/examples/cifar10_pipeline/cifar10_pipeline_interactive.ipynb [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "is_executing": false
+    }
+   },
+   "outputs": [],
+   "source": [
+    "infer_schema = SchemaGen(stats=statistics_gen.outputs['output'])\n",
```
[the remark this answers is not in the document]
Comment (valeriano-manassero) [reply tfx/examples/cifar10_pipeline/cifar10_pipeline_interactive.ipynb]: updated, ty

File: tfx/examples/cifar10_pipeline/cifar10_pipeline_interactive.ipynb [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+    "pycharm": {
+     "is_executing": false
+    },
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "trainer = Trainer(\n",
+    "    module_file=_module_file,\n",
+    "    transformed_examples=transform.outputs['transformed_examples'],\n",
```
[the remark this answers is not in the document]
Comment (valeriano-manassero) [reply tfx/examples/cifar10_pipeline/cifar10_pipeline_interactive.ipynb]: updated, ty

File: tfx/examples/cifar10_pipeline/README.md (hunk 5 above)
Comment (valeriano-manassero) [reply tfx/examples/cifar10_pipeline/README.md]: This is tricky. i got this data with some scripts I wrote on the fly. This is because I didn't find a clean way yet to import data from various sources with a custom Executor without the need of building a spcific Docker image that is really something complicated for a newcomer playing with the example. Any suggestion is welcome.

Comment (ruoyu90) [reply tfx/examples/cifar10_pipeline/README.md]: I see. It's ok, we will try to fix that.
